source/solvers/fluid_dynamics_matrix_free_operators.cc

```diff
@@ -1164,24 +1182,56 @@ NavierStokesOperatorBase<dim, number>::do_local_weak_dirichlet_bc(
       auto       value    = integrator.get_value(q);
       const auto gradient = integrator.get_gradient(q);
 
-      // If we are assembling the residual, substract the target velocity from
-      // the velocity value.
-      if constexpr (assemble_residual)
-        for (unsigned int d = 0; d < dim; ++d)
-          value[d] -= this->face_target_velocity[face_index][q][d];
+      if (this->boundary_conditions.type[boundary_index] ==
+          BoundaryConditions::BoundaryType::function_weak)
+        {
+          // If we are assembling the residual, substract the target velocity
+          // from the velocity value.
+          if constexpr (assemble_residual)
+            for (unsigned int d = 0; d < dim; ++d)
+              value[d] -= this->face_target_velocity[face_index][q][d];
 
-      for (unsigned int d = 0; d < dim; ++d)
-        { // Assemble (v,beta (u-u_target)) for the main penalization
-          value_result[d] += penalty_parameter * value[d];
+          for (unsigned int d = 0; d < dim; ++d)
+            { // Assemble (v,beta (u-u_target)) for the main penalization
+              value_result[d] += penalty_parameter * value[d];
 
-          // Assemble ν(v,∇δu·n)
-          for (unsigned int i = 0; i < dim; ++i)
-            value_result[d] -=
-              kinematic_viscosity * gradient[d][i] * normal_vector[i];
-          // Assemble ν(∇v·n,(u-u_target))
-          for (unsigned int i = 0; i < dim; ++i)
-            gradient_result[d][i] -=
-              kinematic_viscosity * value[d] * normal_vector[i];
+              // Assemble ν(v,∇δu·n)
+              for (unsigned int i = 0; i < dim; ++i)
+                value_result[d] -=
+                  kinematic_viscosity * gradient[d][i] * normal_vector[i];
+              // Assemble ν(∇v·n,(u-u_target))
+              for (unsigned int i = 0; i < dim; ++i)
+                gradient_result[d][i] -=
+                  kinematic_viscosity * value[d] * normal_vector[i];
+            }
+        }
+      else if (this->boundary_conditions.type[boundary_index] ==
+               BoundaryConditions::BoundaryType::outlet)
```

Comment:
I would restructure the code a bit, since gradients are not needed in this code path.

Reply:
True, I will refactor the code a bit to make sure that we only do what the specific boundary condition requires